edb/edgeql/compiler/func.py

```diff
@@ -383,6 +383,24 @@ def compile_operator(
                 hint = (f"The IF and ELSE result clauses must be of "
                         f"compatible types, while the condition clause must "
                         f"be 'std::bool'. {hint}")
+            elif op_name == '+':
+                str_t = cast(s_scalars.ScalarType,
+                             env.schema.get('std::str'))
+                bytes_t = cast(s_scalars.ScalarType,
+                               env.schema.get('std::bytes'))
+                if (
+                    (ltype.issubclass(env.schema, str_t) and
+                        rtype.issubclass(env.schema, str_t)) or
+                    (ltype.issubclass(env.schema, bytes_t) and
+                        rtype.issubclass(env.schema, bytes_t)) or
+                    (ltype.is_array() and rtype.is_array())
+                ):
+                    raise errors.QueryError(
+                        f"operator '+' cannot be applied to "
+                        f"operands of type {types}; "
+                        f"use the '++' operator instead",
```

Comment:
I feel like `"use the '++' operator instead"` logically belongs to the `hint`

Reply:
It does, but what if the hint isn't rendered? A hint is something that might not actually be the right thing in many cases; it's our attempt to suggest a solution. In this case we know the only right solution -- use "++". So I'd keep it right in the error message.

Reply:
https://stratechery.com/2019/aws-mongodb-and-the-economic-realities-of-open-source/

Reply:
I think hints are good when we have an idea of how a user can fix their stuff, but not absolutely sure. Besides, hints might not be rendered in all environments. In this case we know that the only way to fix the error is using "++", it's simple. So I'd keep this right in the error message.

Reply:
I disagree. The purpose of a hint is exactly for this: point users to a way to fix the error.  The idea is that hint and/or details are always rendered when the error is user-facing, e.g in repl.

Reply:
You seem to disagree stronger so I'll update the code to use hints. :)